### PR TITLE
HPCC-12267 Initial implementation of QUANTILE activity in roxie

### PR DESCRIPTION
**Pull request by ghalliday**

### CODE REVIEW

**Comment by ghalliday** [comment]
@richardkchapman @jamienoss  please review.

**Comment by richardkchapman** [comment]
The change to alreadySorted could have interesting implications for other activities/existing code, I think? Did you check regression suite?

File: ecl/hql/hqlmeta.cpp [hunk 1]
```diff
@@ -2443,6 +2443,22 @@ void calculateDatasetMeta(CHqlMetaInfo & meta, IHqlExpression * expr)
             }
             break;
         }
+    case no_quantile:
+        //It is not be safe to assume that all implementations of the quantile activity will generate their results in order.
```
Comment (richardkchapman) [inline ecl/hql/hqlmeta.cpp]: Grammar: "It is not be ..."

File: plugins/cassandra/cpp-driver [hunk 2]
```diff
@@ -1 +1 @@
-Subproject commit d7bbad34db39a51f209c6fadd07c9ec3a0bb86b7
```
Comment (richardkchapman) [inline plugins/cassandra/cpp-driver]: Unwanted and unrelated change

File: rtl/include/eclhelper_base.hpp [hunk 3]
```diff
@@ -708,6 +707,30 @@ class CThorProjectArg : public CThorArg, implements IHThorProjectArg
...
+        {
+        case TAIarg:
+        case TAIquantilearg_1:
+            return static_cast<IHThorQuantileArg *>(this);
+        }
+        return NULL;
+    }
+
+    virtual unsigned getFlags() { return 0; }
+    virtual unsigned __int64 getNumDivisions() { return 2; }
```
Comment (richardkchapman) [inline rtl/include/eclhelper_base.hpp]: Seems a weird default

**Comment by richardkchapman** [comment]
Minor issues picked up on quick scan. I'll stare a the Roxie bits in more detail in a sec.

File: rtl/include/eclhelper_base.hpp (hunk 3 above)
Comment (ghalliday) [inline rtl/include/eclhelper_base.hpp]: If num divisions is 2 that means it will calculate the median.  (If calculate centiles, then it calculates 99 split points.)

File: plugins/cassandra/cpp-driver (hunk 2 above)
Comment (ghalliday) [inline plugins/cassandra/cpp-driver]: I thought this might be there, but couldn't see which commit it was in - I'll search

**Comment by ghalliday** [comment]
Yes the isSorted change had no affect on the existing regression suite.  It had a minor effect on 2 quantile examples.

**Comment by ghalliday** [comment]
I have fixed the incorrect cassandra change by removing it from the commit.  I'll update with other fixes when review finished.

File: roxie/ccd/ccdserver.cpp [hunk 4]
```diff
@@ -7466,6 +7466,268 @@ IRoxieServerActivityFactory *createRoxieServerSortActivityFactory(unsigned _id,
...
+    bool isQuantileIncluded(unsigned quantile) const
+    {
+        if (quantile == 0)
+            return (flags & TQFfirst) != 0;
+        if (quantile == numDivisions)
+            return (flags & TQFlast) != 0;
+        if (rangeIsAll)
+            return true;
+        //Compare against the list of ranges provided
+        //MORE: Since the list is sorted should only need to compare the next (and allow for dups)
```
Comment (richardkchapman) [inline roxie/ccd/ccdserver.cpp]: Are you planning to address this change before merging?

You could also use bsearch...

File: ecl/hql/hqlattr.cpp [hunk 5]
```diff
@@ -2813,6 +2814,27 @@ IHqlExpression * calcRowInformation(IHqlExpression * expr)
...
+                    parts++;
+
+                IHqlExpression * transform = queryNewColumnProvider(expr);
+                if (transformContainsSkip(transform) || expr->hasAttribute(dedupAtom))
+                    info.setRange(0,parts-1);
+                else
+                    info.setN(parts-1);
+            }
+            else
+                info.setUnknown(RCMfew);
```
Comment (richardkchapman) [inline ecl/hql/hqlattr.cpp]: If it's grouped, isn't the expected rowcount some fraction of the input's rowcount (i.e. not necessarily few) ?

File: ecl/hqlcpp/hqlhtcpp.cpp [hunk 6]
```diff
@@ -16584,6 +16587,99 @@ ABoundActivity * HqlCppTranslator::doBuildActivitySort(BuildCtx & ctx, IHqlExpre
...
+    IHqlExpression * number = expr->queryChild(1);
+    IHqlExpression * sortlist = expr->queryChild(2);
+    IHqlExpression * transform = expr->queryChild(3);
+    IHqlExpression * score = queryAttributeChild(expr, scoreAtom, 0);
+    IHqlExpression * skew = queryAttributeChild(expr, skewAtom, 0);
+    IHqlExpression * dedupAttr = expr->queryAttribute(dedupAtom);
+    IHqlExpression * range = queryAttributeChild(expr, rangeAtom, 0);
+
+    instance->classctx.addQuotedLiteral("virtual ICompare * queryCompare() { return &compare; }");
+
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlhtcpp.cpp]: Why all the blank lines?

File: ecl/hqlcpp/hqlhtcpp.cpp [hunk 7]
```diff
@@ -16584,6 +16587,99 @@ ABoundActivity * HqlCppTranslator::doBuildActivitySort(BuildCtx & ctx, IHqlExpre
...
+        BuildCtx funcctx(instance->startctx);
+        funcctx.addQuotedCompound("virtual unsigned __int64 getScore(const void * _self)");
+        funcctx.addQuotedLiteral("unsigned char * self = (unsigned char *) _self;");
+        bindTableCursor(funcctx, dataset, "self");
+        buildReturn(funcctx, score);
+    }
+
+    IHqlExpression * counter = queryAttributeChild(expr, _countProject_Atom, 0);
+    IHqlExpression * selSeq = querySelSeq(expr);
+    {
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlhtcpp.cpp]: Should this be conditional on the needsTransform flag and/or transform != NULL?

File: ecl/hqlcpp/hqlhtcpp.cpp [hunk 8]
```diff
@@ -16584,6 +16587,99 @@ ABoundActivity * HqlCppTranslator::doBuildActivitySort(BuildCtx & ctx, IHqlExpre
...
+    if (range)
+        flags.append("|TQFhasrange");
+    if (skew)
+        flags.append("|TQFhasskew");
+    if (dedupAttr)
+        flags.append("|TQFdedup");
+    if (unstable)
+        flags.append("|TQFunstable");
+    if (!number->queryValue())
+        flags.append("|TQFvariabledivisions");
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlhtcpp.cpp]: Roxie implementation does not use this flag - is that correct? I assume it can't vary per group, only per instance of a child query?

**Comment by richardkchapman** [comment]
I'm done reviewing. Looks good to me, minor questions only. @jamienoss over to you.

File: ecl/hql/hqlattr.cpp [hunk 9]
```diff
@@ -629,7 +630,7 @@ unsigned getOperatorMetaFlags(node_operator op)
     case no_unused40: case no_unused41: case no_unused42: case no_unused43: case no_unused44: case no_unused45: case no_unused46: case no_unused47: case no_unused48: case no_unused49:
     case no_unused50: case no_unused52:
     case no_unused80: case no_unused83:
-    case no_unused101: case no_unused102:
+    case no_unused102:
```
Comment (jamienoss) [inline ecl/hql/hqlattr.cpp]: Why consume 101?

File: ecl/hql/hqlgram2.cpp [hunk 10]
```diff
@@ -5770,6 +5770,7 @@ IHqlExpression * HqlGram::processSortList(const attribute & errpos, node_operato
     return NULL;
 }
 
+
```
Comment (jamienoss) [inline ecl/hql/hqlgram2.cpp]: newline?

File: ecl/hql/hqlmeta.cpp [hunk 11]
```diff
@@ -1540,13 +1540,13 @@ static IHqlExpression * createSubSorted(IHqlExpression * dataset, IHqlExpression
...
 
-    assertex(isAlreadySorted(subsort, order, isLocal||alwaysLocal, ignoreGrouping));
+    assertex(isAlreadySorted(subsort, order, isLocal||alwaysLocal, ignoreGrouping, false));
     return subsort.getClear();
 }
 
 IHqlExpression * getSubSort(IHqlExpression * dataset, const HqlExprArray & order, bool isLocal, bool ignoreGrouping, bool alwaysLocal)
 {
-    if (isAlreadySorted(dataset, order, isLocal||alwaysLocal, ignoreGrouping))
+    if (isAlreadySorted(dataset, order, isLocal||alwaysLocal, ignoreGrouping, true))  // could possible have requireDistribution = false
```
Comment (jamienoss) [inline ecl/hql/hqlmeta.cpp]: possible -> possibly
Would it be a good idea to expand on this comment and state why?

File: ecl/hql/hqlmeta.cpp [hunk 12]
```diff
@@ -1556,7 +1556,7 @@ IHqlExpression * getSubSort(IHqlExpression * dataset, const HqlExprArray & order
 
 IHqlExpression * getSubSort(IHqlExpression * dataset, IHqlExpression * order, bool isLocal, bool ignoreGrouping, bool alwaysLocal)
 {
-    if (isAlreadySorted(dataset, order, isLocal||alwaysLocal, ignoreGrouping))
+    if (isAlreadySorted(dataset, order, isLocal||alwaysLocal, ignoreGrouping, true))  // could possible have requireDistribution = false
```
Comment (jamienoss) [inline ecl/hql/hqlmeta.cpp]: possible -> possibly

File: ecl/hql/hqlmeta.cpp [hunk 13]
```diff
@@ -2443,6 +2443,22 @@ void calculateDatasetMeta(CHqlMetaInfo & meta, IHqlExpression * expr)
...
+            OwnedHqlExpr leftSelect = createSelector(no_left, dataset, expr->queryAttribute(_selectorSequence_Atom));
+            TableProjectMapper mapper;
+            mapper.setMapping(transform, leftSelect);
+
+            extractMeta(meta, dataset);
+            meta.removeAllSortOrders();
+            meta.applyProject(mapper);  // distribution preserved if local..
+        }
+        else
+            meta.preserveGrouping(dataset);
```
Comment (jamienoss) [inline ecl/hql/hqlmeta.cpp]: Something seemed odd (semanticly) given the previous isGrouped(dataset), looking, I see that preserveGrouping also checks this.

In addition CHqlMetaInfo::clearGrouping() also clears groupSortOrder, should preserveGrouping do so also and thus just call clearGrouping()?

File: ecl/hqlcpp/hqlhtcpp.cpp (hunk 7 above)
Comment (jamienoss) [inline ecl/hqlcpp/hqlhtcpp.cpp]: Why is this blocked at all? Missing if statement, per Richards comment above?

File: ecl/regress/quantile_e1.ecl [hunk 14]
```diff
@@ -0,0 +1,44 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems.
```
Comment (jamienoss) [inline ecl/regress/quantile_e1.ecl]: Missing registered trademark symbol on HPCC Systems - similarly for all other newly added tests etc.

File: testing/regress/ecl/quantile1.ecl [hunk 15]
```diff
@@ -0,0 +1,58 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems.
```
Comment (jamienoss) [inline testing/regress/ecl/quantile1.ecl]: reg trademark.

File: testing/regress/ecl/quantile10.ecl [hunk 16]
```diff
@@ -0,0 +1,69 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems.
```
Comment (jamienoss) [inline testing/regress/ecl/quantile10.ecl]: reg trademark.

File: testing/regress/ecl/quantile11.ecl [hunk 17]
```diff
@@ -0,0 +1,47 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems.
```
Comment (jamienoss) [inline testing/regress/ecl/quantile11.ecl]: reg trademark.

File: testing/regress/ecl/quantile12.ecl [hunk 18]
```diff
@@ -0,0 +1,59 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems.
```
Comment (jamienoss) [inline testing/regress/ecl/quantile12.ecl]: reg trademark.

File: testing/regress/ecl/quantile13.ecl [hunk 19]
```diff
@@ -0,0 +1,50 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems.
```
Comment (jamienoss) [inline testing/regress/ecl/quantile13.ecl]: reg trademark.

File: testing/regress/ecl/quantile1.ecl [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+show(QUANTILE(ascending, 10, {id}, createQuantile(LEFT, COUNTER), FIRST, LAST));        // 0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 99
+show(QUANTILE(descending, 10, {id}, createQuantile(LEFT, COUNTER), FIRST, LAST));        // 0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 99
+show(QUANTILE(ascendingBy3, 10, {id}, createQuantile(LEFT, COUNTER), FIRST, LAST));        // 0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 99
+
+//Calculate median
+show(QUANTILE(ascending, 2, {id}, createQuantile(LEFT, COUNTER)));        // 50
+show(QUANTILE(descending, 2, {id}, createQuantile(LEFT, COUNTER)));        // 50
+show(QUANTILE(ascendingBy3, 2, {id}, createQuantile(LEFT, COUNTER)));        // 50
+
+//Check that first/last work with median - may need to fall back to simplest implementation
```
Comment (jamienoss) [inline testing/regress/ecl/quantile1.ecl]: Should the latter half of this comment be present?

**Comment by jamienoss** [comment]
@ghalliday Few comments - back to you.
@richardkchapman just checking over the actual roxie activity code now.
